File: Algorithm/Python/451. Sort Characters By Frequency.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-12-23 16:43:34
# @Last modified by:   WuLC
# @Last Modified time: 2016-12-23 17:03:35
# @Email: liangchaowu5@gmail.com


# sort the hashmap
class Solution(object):
    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        """
        result = ''
        count = {}
        for char in s:
            count.setdefault(char, 0)
            count[char] += 1
        sorted_dict = sorted(count.items(), key = lambda x:x[1], reverse = True)
        for pair in sorted_dict:
            result += pair[0]*pair[1]
        return result

File: Algorithm/Java/451. Sort Characters By Frequency.java
/*
* @Author: WuLC
* @Date:   2017-07-04 14:55:50
* @Last Modified by:   WuLC
* @Last Modified time: 2017-07-04 14:56:23
* @Email: liangchaowu5@gmail.com
*/

// priority queue
public class Solution 
{
    public String frequencySort(String s) 
    {
        Map<Character, Integer> count = new HashMap<Character, Integer>();
        for (char c : s.toCharArray())
        {
            if (!count.containsKey(c)) count.put(c, 0);
            count.put(c, count.get(c)+1);
        }
        
        PriorityQueue<Map.Entry<Character, Integer>> pq = new PriorityQueue<Map.Entry<Character, Integer>>(
            new Comparator<Map.Entry<Character, Integer>>()
            {
                @Override
                public int compare(Map.Entry<Character, Integer> m1, Map.Entry<Character, Integer> m2)
                {
                    return m2.getValue() - m1.getValue();
                }
            }
        );
        
        pq.addAll(count.entrySet());
        String result = "";
        Map.Entry<Character, Integer> tmp;
        while(pq.size() > 0)
        {
            tmp = pq.poll();
            result += new String(new char[tmp.getValue()]).replace("\0", String.valueOf(tmp.getKey()));
        }
        return result;
    }
}